File: xlox/java/src/com/cuongd/nanosystems/xlox/Scanner.java
package com.cuongd.nanosystems.xlox;

import static com.cuongd.nanosystems.xlox.TokenType.AND;
import static com.cuongd.nanosystems.xlox.TokenType.BANG;
import static com.cuongd.nanosystems.xlox.TokenType.BANG_EQUAL;
import static com.cuongd.nanosystems.xlox.TokenType.BREAK;
import static com.cuongd.nanosystems.xlox.TokenType.CLASS;
import static com.cuongd.nanosystems.xlox.TokenType.COLON;
import static com.cuongd.nanosystems.xlox.TokenType.COMMA;
import static com.cuongd.nanosystems.xlox.TokenType.DOT;
import static com.cuongd.nanosystems.xlox.TokenType.ELSE;
import static com.cuongd.nanosystems.xlox.TokenType.EOF;
import static com.cuongd.nanosystems.xlox.TokenType.EQUAL;
import static com.cuongd.nanosystems.xlox.TokenType.EQUAL_EQUAL;
import static com.cuongd.nanosystems.xlox.TokenType.FALSE;
import static com.cuongd.nanosystems.xlox.TokenType.FOR;
import static com.cuongd.nanosystems.xlox.TokenType.FUN;
import static com.cuongd.nanosystems.xlox.TokenType.GREATER;
import static com.cuongd.nanosystems.xlox.TokenType.GREATER_EQUAL;
import static com.cuongd.nanosystems.xlox.TokenType.IDENTIFIER;
import static com.cuongd.nanosystems.xlox.TokenType.IF;
import static com.cuongd.nanosystems.xlox.TokenType.LEFT_BRACE;
import static com.cuongd.nanosystems.xlox.TokenType.LEFT_PAREN;
import static com.cuongd.nanosystems.xlox.TokenType.LESS;
import static com.cuongd.nanosystems.xlox.TokenType.LESS_EQUAL;
import static com.cuongd.nanosystems.xlox.TokenType.MINUS;
import static com.cuongd.nanosystems.xlox.TokenType.NIL;
import static com.cuongd.nanosystems.xlox.TokenType.NUMBER;
import static com.cuongd.nanosystems.xlox.TokenType.OR;
import static com.cuongd.nanosystems.xlox.TokenType.PLUS;
import static com.cuongd.nanosystems.xlox.TokenType.PRINT;
import static com.cuongd.nanosystems.xlox.TokenType.QUESTION;
import static com.cuongd.nanosystems.xlox.TokenType.RETURN;
import static com.cuongd.nanosystems.xlox.TokenType.RIGHT_BRACE;
import static com.cuongd.nanosystems.xlox.TokenType.RIGHT_PAREN;
import static com.cuongd.nanosystems.xlox.TokenType.SEMICOLON;
import static com.cuongd.nanosystems.xlox.TokenType.SLASH;
import static com.cuongd.nanosystems.xlox.TokenType.STAR;
import static com.cuongd.nanosystems.xlox.TokenType.STRING;
import static com.cuongd.nanosystems.xlox.TokenType.SUPER;
import static com.cuongd.nanosystems.xlox.TokenType.THIS;
import static com.cuongd.nanosystems.xlox.TokenType.TRUE;
import static com.cuongd.nanosystems.xlox.TokenType.VAR;
import static com.cuongd.nanosystems.xlox.TokenType.WHILE;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Scanner {
  private final String source;
  private final List<Token> tokens = new ArrayList<>();
  private int start = 0;
  private int current = 0;
  private int line = 1;

  private static final Map<String, TokenType> keywords;

  static {
    keywords = new HashMap<>();
    keywords.put("and", AND);
    keywords.put("break", BREAK);
    keywords.put("class", CLASS);
    keywords.put("else", ELSE);
    keywords.put("false", FALSE);
    keywords.put("for", FOR);
    keywords.put("fun", FUN);
    keywords.put("if", IF);
    keywords.put("nil", NIL);
    keywords.put("or", OR);
    keywords.put("print", PRINT);
    keywords.put("return", RETURN);
    keywords.put("super", SUPER);
    keywords.put("this", THIS);
    keywords.put("true", TRUE);
    keywords.put("var", VAR);
    keywords.put("while", WHILE);
  }

  Scanner(String source) {
    this.source = source;
  }

  List<Token> scanTokens() {
    while (!isAtEnd()) {
      start = current;
      scanToken();
    }

    tokens.add(new Token(EOF, "", null, line));
    return tokens;
  }

  private void scanToken() {
    char c = advance();
    switch (c) {
      case '(':
        addToken(LEFT_PAREN);
        break;
      case ')':
        addToken(RIGHT_PAREN);
        break;
      case '{':
        addToken(LEFT_BRACE);
        break;
      case '}':
        addToken(RIGHT_BRACE);
        break;
      case ',':
        addToken(COMMA);
        break;
      case '.':
        addToken(DOT);
        break;
      case '-':
        addToken(MINUS);
        break;
      case '+':
        addToken(PLUS);
        break;
      case ';':
        addToken(SEMICOLON);
        break;
      case '*':
        addToken(STAR);
        break;
      case '?':
        addToken(QUESTION);
        break;
      case ':':
        addToken(COLON);
        break;

      case '!':
        addToken(match('=') ? BANG_EQUAL : BANG);
        break;
      case '=':
        addToken(match('=') ? EQUAL_EQUAL : EQUAL);
        break;
      case '>':
        addToken(match('=') ? GREATER_EQUAL : GREATER);
        break;
      case '<':
        addToken(match('=') ? LESS_EQUAL : LESS);
        break;
      case '/':
        if (match('/')) {
          while (peek() != '\n' && !isAtEnd()) advance();
        } else if (match('*')) {
          int nesting = 1;
          while (nesting > 0) {
            if (isAtEnd()) {
              XLox.error(nesting, "Unterminated block comment.");
              return;
            }
            if (peek() == '/' && peekNext() == '*') {
              nesting++;
              advance();
              advance();
              continue;
            }
            if (peek() == '*' && peekNext() == '/') {
              nesting--;
              advance();
              advance();
              continue;
            }
            if (peek() == '\n') line++;
            advance();
          }
        } else {
          addToken(SLASH);
        }
        break;

      case ' ':
      case '\r':
      case '\t':
        break;
      case '\n':
        line++;
        break;

      case '"':
        string();
        break;
      default:
        if (isDigit(c)) {
          number();
        } else if (isAlpha(c)) {
          identifier();
        } else {
          XLox.error(line, "Unexpected character.");
        }
        break;
    }
  }

  private void string() {
    while (peek() != '"' && !isAtEnd()) {
      if (peek() == '\n') line++;
      advance();
    }

    if (isAtEnd()) {
      XLox.error(line, "Unterminated string.");
      return;
    }

    // The closing ".
    advance();

    String value = source.substring(start + 1, current - 1);
    addToken(STRING, value);
  }

  private void number() {
    while (isDigit(peek())) advance();

    if (peek() == '.' && isDigit(peekNext())) {
      // Consume the '.'.
      advance();
      while (isDigit(peek())) advance();
    }

    Double value = Double.parseDouble(source.substring(start, current));
    addToken(NUMBER, value);
  }

  private void identifier() {
    while (isAlphaNumeric(peek())) advance();

    String text = source.substring(start, current);
    TokenType type = keywords.getOrDefault(text, IDENTIFIER);
    addToken(type);
  }

  private boolean isDigit(char c) {
    return c >= '0' && c <= '9';
  }

  private boolean isAlpha(char c) {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c == '_');
  }

  private boolean isAlphaNumeric(char c) {
    return isAlpha(c) || isDigit(c);
  }

  private boolean match(char expected) {
    if (isAtEnd()) return false;
    if (source.charAt(current) != expected) return false;

    advance();
    return true;
  }

  private boolean isAtEnd() {
    return current >= source.length();
  }

  private char advance() {
    return source.charAt(current++);
  }

  private char peek() {
    if (isAtEnd()) return '\0';
    return source.charAt(current);
  }

  private char peekNext() {
    if (current + 1 >= source.length()) return '\0';
    return source.charAt(current + 1);
  }

  private void addToken(TokenType type, Object literal) {
    String text = source.substring(start, current);
    tokens.add(new Token(type, text, literal, line));
  }

  private void addToken(TokenType type) {
    addToken(type, null);
  }
}

File: zlox/scanner.zig
const std = @import("std");

pub const Scanner = struct {
    source: []const u8,
    start: usize,
    current: usize,
    line: usize,

    pub fn init(source: []const u8) Scanner {
        return .{ .source = source, .start = 0, .current = 0, .line = 1 };
    }

    pub fn scanToken(self: *Scanner) Token {
        self.skipWhitespace();
        self.start = self.current;
        if (self.isAtEnd()) return self.makeToken(.EOF);

        const c = self.peek();
        self.advance();

        if (isAlpha(c)) return self.identifier();
        if (std.ascii.isDigit(c)) return self.number();

        switch (c) {
            '(' => return self.makeToken(.LEFT_PAREN),
            ')' => return self.makeToken(.RIGHT_PAREN),
            '{' => return self.makeToken(.LEFT_BRACE),
            '}' => return self.makeToken(.RIGHT_BRACE),
            ';' => return self.makeToken(.SEMICOLON),
            ',' => return self.makeToken(.COMMA),
            '.' => return self.makeToken(.DOT),
            '-' => return self.makeToken(.MINUS),
            '+' => return self.makeToken(.PLUS),
            '/' => return self.makeToken(.SLASH),
            '*' => return self.makeToken(.STAR),

            '!' => return self.makeToken(if (self.match('=')) .BANG_EQUAL else .BANG),
            '=' => return self.makeToken(if (self.match('=')) .EQUAL_EQUAL else .EQUAL),
            '<' => return self.makeToken(if (self.match('=')) .LESS_EQUAL else .LESS),
            '>' => return self.makeToken(if (self.match('=')) .GREATER_EQUAL else .GREATER),

            '"' => return self.string(),

            else => return self.errorToken("Unexpected character."),
        }
    }

    fn isAtEnd(self: *const Scanner) bool {
        return self.current == self.source.len;
    }

    fn advance(self: *Scanner) void {
        self.current += 1;
    }

    fn peek(self: *const Scanner) u8 {
        return if (self.isAtEnd()) 0 else self.source[self.current];
    }

    fn peekNext(self: *const Scanner) u8 {
        return if (self.current + 1 == self.source.len) 0 else self.source[self.current + 1];
    }

    fn match(self: *Scanner, c: u8) bool {
        if (self.isAtEnd()) return false;
        if (self.source[self.current] != c) return false;
        self.advance();
        return true;
    }

    fn makeToken(self: *const Scanner, tokenType: TokenType) Token {
        return .{ .type = tokenType, .lexeme = self.source[self.start..self.current], .line = self.line };
    }

    fn errorToken(self: *const Scanner, message: []const u8) Token {
        return .{ .type = .ERROR, .lexeme = message, .line = self.line };
    }

    fn skipWhitespace(self: *Scanner) void {
        while (true) {
            const c = self.peek();
            switch (c) {
                ' ', '\r', '\t' => self.advance(),

                '\n' => {
                    self.line += 1;
                    self.advance();
                },

                '/' => if (self.peekNext() == '/') {
                    while (self.peek() != '\n' and !self.isAtEnd()) self.advance();
                } else {
                    return;
                },

                else => return,
            }
        }
    }

    fn identifier(self: *Scanner) Token {
        while (isAlpha(self.peek()) or std.ascii.isDigit(self.peek())) self.advance();

        const identifierType: TokenType = switch (self.source[self.start]) {
            'a' => self.checkKeyword(1, "nd", .AND),
            'c' => self.checkKeyword(1, "lass", .CLASS),
            'e' => self.checkKeyword(1, "lse", .ELSE),
            'i' => self.checkKeyword(1, "f", .IF),
            'n' => self.checkKeyword(1, "il", .NIL),
            'o' => self.checkKeyword(1, "r", .OR),
            'p' => self.checkKeyword(1, "rint", .PRINT),
            'r' => self.checkKeyword(1, "eturn", .RETURN),
            's' => self.checkKeyword(1, "uper", .SUPER),
            'v' => self.checkKeyword(1, "ar", .VAR),
            'w' => self.checkKeyword(1, "hile", .WHILE),

            'f' => if (self.current - self.start > 1)
                switch (self.source[self.start + 1]) {
                    'a' => self.checkKeyword(2, "lse", .FALSE),
                    'o' => self.checkKeyword(2, "r", .FOR),
                    'u' => self.checkKeyword(2, "n", .FUN),
                    else => .IDENTIFIER,
                }
            else
                .IDENTIFIER,

            't' => if (self.current - self.start > 1)
                switch (self.source[self.start + 1]) {
                    'h' => self.checkKeyword(2, "is", .THIS),
                    'r' => self.checkKeyword(2, "ue", .TRUE),
                    else => .IDENTIFIER,
                }
            else
                .IDENTIFIER,
            else => .IDENTIFIER,
        };

        return self.makeToken(identifierType);
    }

    fn checkKeyword(self: *const Scanner, start: usize, rest: []const u8, tokenType: TokenType) TokenType {
        if (self.current - self.start - start == rest.len and
            std.mem.eql(u8, self.source[self.start + start .. self.current], rest))
        {
            return tokenType;
        }
        return .IDENTIFIER;
    }

    fn number(self: *Scanner) Token {
        while (std.ascii.isDigit(self.peek())) self.advance();

        if (self.peek() == '.' and std.ascii.isDigit(self.peekNext())) {
            self.advance();
            while (std.ascii.isDigit(self.peek())) self.advance();
        }

        return self.makeToken(.NUMBER);
    }

    fn string(self: *Scanner) Token {
        while (self.peek() != '"' and !self.isAtEnd()) {
            if (self.peek() == '\n') self.line += 1;
            self.advance();
        }

        if (self.isAtEnd()) return self.errorToken("Unterminated string.");

        // The closing quote.
        self.advance();
        return self.makeToken(.STRING);
    }
};

pub const Token = struct {
    type: TokenType,
    lexeme: []const u8,
    line: usize,
};

pub const TokenType = enum {
    // Single-character tokens.
    LEFT_PAREN,
    RIGHT_PAREN,
    LEFT_BRACE,
    RIGHT_BRACE,
    SEMICOLON,
    COMMA,
    DOT,
    MINUS,
    PLUS,
    SLASH,
    STAR,

    // One or two character tokens.
    BANG,
    BANG_EQUAL,
    EQUAL,
    EQUAL_EQUAL,
    GREATER,
    GREATER_EQUAL,
    LESS,
    LESS_EQUAL,

    // Literals.
    IDENTIFIER,
    STRING,
    NUMBER,

    // Keywords.
    AND,
    CLASS,
    ELSE,
    FALSE,
    FOR,
    FUN,
    IF,
    NIL,
    OR,
    PRINT,
    RETURN,
    SUPER,
    THIS,
    TRUE,
    VAR,
    WHILE,

    ERROR,
    EOF,
};

fn isAlpha(c: u8) bool {
    return std.ascii.isAlphabetic(c) or c == '_';
}
